rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,345 @@
+  string UserMessage = 3 [(gogoproto.jsontag) = "message,omitempty"];
+}
+
+// DatabaseSessionCommandResult represents the result of a user command. It is
+// expected that for each user command/query there will be a corresponding
+// result.
+message DatabaseSessionCommandResult {
+  // Metadata is a common event metadata.
+  Metadata Metadata = 1;
+  // User is a common user event metadata.
+  UserMetadata User = 2;
+  // SessionMetadata is a common event session metadata.
+  SessionMetadata Session = 3;
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 4;
+  // Status of the execution.
+  Status status = 5;
+  // AfftectedRecords represents the number of records that were affected by the
+  // user query.
+  uint64 AffectedRecords = 6;
+}
+```
+
+#### Examples
+
+* Audit event generated for a query that returns data (`SELET id, name FROM events;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true
+  },
+  "affected_records": 5
+}
+```
+
+* Query that only returns a success message (`UPDATE events SET ...`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true
+  },
+  "affected_records": 3
+}
+```
+
+* Query that returns an error (`SELECT err;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": false,
+    "error": "ERROR: column \"err\" does not exist"
+  },
+  "affected_records": 0
+}
+```
+
+### Recording options
+
+Database session recording options will be defined per role. This configuration
+will be under the `options.record_session.db` role option. It will support the
+following values:
+- `on`: Enables session recording.
+- `off`: Disables session recording. Audit events are kept unchanged (start,
+  end, and query events emitted).
+
+The value will be set to `on` by default, keeping the current behavior.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: alice
+spec:
+  allow:
+    ...
+  options:
+    record_session:
+      db: on|off
+```
+
+### Player
+
+To play the session recording, we will rely on the existing player used by SSH
+sessions. This means we don't need to introduce any new player on `tsh` or the
+Web UI. We're going to convert database recording events into `SessionPrint`
+events, which can be rendered by the players.
+
+This conversion will be agnostic to the database protocol. This will make it
+easier to keep the session recording play consistent across protocols. However,
+protocol-specific messages will still require special handling, otherwise the
+player won't be able to present them.
+
+It will present the session recording events in different text formats:
+
+- Status (result): Show the user message. If none is present, show "SUCCESS" in
+  case of success or "ERROR" in case of error.
+- The player will define if there is need to display the number of affected
+  records. Protocol-specific variations can be added for this. For example, if
+  the query was a `SELECT` the player might display it as "Returned X rows"
+  after the status.
```

Comment:
The `DatabaseSessionCommandResult` only captures a number `affected_records`. How does the player known how to render `Returned xxx` vs `xxx inserted`? If the logic is based on the query, how reliable is our logic?